docs/tutorial/writing-tests/test-case.rst

```diff
@@ -0,0 +1,1105 @@
+    100  5105    0  5105    0     0   6645      0 --:--:-- --:--:-- --:--:--  6647
+    ------------------------------------------------------------------------- >8 ---
+    Outcome: job passed
+    Finalizing session that hasn't been submitted anywhere: checkbox-run-2024-08-06T14.17.23
+    ==================================[ Results ]===================================
+     ☑ : Test that the network speed is acceptable
+
+
+.. warning::
+
+    Don't assume that a Checkbox job will inherit any environment variable from
+    the parent shell, global env or any other source. There are a few exceptions
+    but in general:
+
+    - Any variable that is not in the ``environ`` section of a job is not set
+    - Any variable not declared in the ``environment`` section of a launcher or configuration file is not set
+
+If you decide to parametrize your tests using environment variables, always
+check if they are set or give them a default value via ``${...:-default}``.
+If you expect a variable to be set and it is not, always fail the test stating
+what variable you needed and what it was for. If you decide to use a default
+value, always output the value the test is going to use in the test log so that
+when you have to investigate why something went wrong, it is trivial to
+reproduce the tests with the parameters that may have made it fail.
+
+Resources
+=========
+
+Before even thinking to test if we are connected to the internet a wise
+question to ask would be: do we even have a network interface? Lets create a
+resource job to fetch this information.
+
+Create a new job with the following content:
+
+.. code-block:: none
+
+    id: network_iface_info
+    _summary: Fetches information of all network intefaces
+    plugin: resource
+    command:
+      ip -details -json link show | jq -r '
+          .[] | "interface: " + .ifname +
+          "\nlink_info_kind: " + .linkinfo.info_kind +
+          "\nlink_type: " + .link_type + "\n"'
+
+This test adds a new dependency to our provider. We need to declare this in
+the correct spot else this will not work in a reproducible manner. Lets create
+a packaging meta-data unit.
+
+.. code-block:: none
+
+    id: extended_tutorial_dependencies
+    unit: packaging meta-data
+    os-id: debian
+    Depends:
+      jq
+
+If you now run the following command you will notice a validation error.
+
+.. code-block:: none
+
+
+    (checkbox_venv) > python3 manage.py validate
+    [...]
+    error: ../base/units/submission/packaging.pxu:3: field 'Depends', clashes with 1 other unit, look at: ../base/units/submission/packaging.pxu:1-3, units/extended_tutorial.pxu:1-4
+    Validation of provider tutorial has failed
+
+Opening the file that the validator complains about, you will notice that the
+jq dependency is already required by a base provider test. We can rely on the
+base provider, so we can safely remove this dependency from our provider.
```

Comment:
This gets difficult pretty quickly, mixing packaging meta data units with resource jobs. Maybe this can be skipped for now, given that the base provider includes it (and, actually, it's not required to run in a venv)?

Reply:
I tried to give a progression that goes through the path of a new job developer in a rational progression. I think that here one would ask themselves: well I've used this tool, but how does checkbox know?